File: internal/repository/groupRepository.go
package repository

import (
	"database/sql"
	"log"

	"Student-management-system/internal/models"
)

type GroupRepository struct {
	db *sql.DB
}

func NewGroupRepository(db *sql.DB) *GroupRepository {
	return &GroupRepository{db: db}
}

func (r *GroupRepository) GetByID(id int) (*models.Group, error) {
	query := "SELECT id, subdivision_id, name, year FROM groups WHERE id = $1"
	row := r.db.QueryRow(query, id)

	var group models.Group
	err := row.Scan(&group.ID, &group.SubdivisionID, &group.Name, &group.Year)
	if err != nil {
		log.Printf("Ошибка при получении группы: %v", err)
		return nil, err
	}

	return &group, nil
}

func (r *GroupRepository) Create(group *models.Group) error {
	query := "INSERT INTO groups (subdivision_id, name, year) VALUES ($1, $2, $3) RETURNING id"
	err := r.db.QueryRow(query, group.SubdivisionID, group.Name, group.Year).Scan(&group.ID)
	if err != nil {
		log.Printf("Ошибка создания группы: %v", err)
		return err
	}

	return nil
}

func (r *GroupRepository) Update(group *models.Group) error {
	query := "UPDATE groups SET subdivision_id = $1, name = $2, year = $3 WHERE id = $4"
	_, err := r.db.Exec(query, group.SubdivisionID, group.Name, group.Year, group.ID)
	return err
}

func (r *GroupRepository) Delete(id int) error {
	query := "DELETE FROM groups WHERE id = $1"
	_, err := r.db.Exec(query, id)
	return err
}

File: app/repositories/group_repository.py
from typing import Optional, List, Dict, Any
from asyncpg import Connection
from .base import BaseRepository
from ..models.group import Group, GroupCreate, GroupUpdate, GroupWithStats


class GroupRepository(BaseRepository[Group]):
    """Репозиторий для работы с группами"""
    
    @property
    def table_name(self) -> str:
        return "groups"
    
    @property
    def model_class(self):
        return Group
    
    async def create(self, data: GroupCreate, conn: Optional[Connection] = None) -> Group:
        """Создать группу"""
        query = """
            INSERT INTO groups (subdivisionid, name, year) 
            VALUES ($1, $2, $3) 
            RETURNING *
        """
        
        async with self._get_connection(conn) as connection:
            row = await connection.fetchrow(
                query, 
                data.subdivision_id, 
                data.name, 
                data.year
            )
            return Group(**dict(row))
    
    async def update(self, id: int, data: GroupUpdate, conn: Optional[Connection] = None) -> Optional[Group]:
        """Обновить группу"""
        update_data = data.model_dump(exclude_unset=True)
        if not update_data:
            return await self.get_by_id(id, conn)
        
        # Маппинг имен полей
        field_mapping = {
            'subdivision_id': 'subdivisionid'
        }
        
        set_parts = []
        values = [id]
        for i, (field, value) in enumerate(update_data.items()):
            db_field = field_mapping.get(field, field)
            set_parts.append(f"{db_field} = ${i+2}")
            values.append(value)
        
        query = f"""
            UPDATE groups 
            SET {', '.join(set_parts)}
            WHERE id = $1
            RETURNING *
        """
        
        async with self._get_connection(conn) as connection:
            row = await connection.fetchrow(query, *values)
            return Group(**dict(row)) if row else None
    
    async def get_by_name(self, name: str, conn: Optional[Connection] = None) -> Optional[Group]:
        """Получить группу по имени"""
        query = "SELECT * FROM groups WHERE name = $1"
        
        async with self._get_connection(conn) as connection:
            row = await connection.fetchrow(query, name)
            return Group(**dict(row)) if row else None
    
    async def get_by_subdivision(self, subdivision_id: int, year: Optional[int] = None, conn: Optional[Connection] = None) -> List[Group]:
        """Получить группы подразделения"""
        if year:
            query = "SELECT * FROM groups WHERE subdivisionid = $1 AND year = $2 ORDER BY name"
            params = [subdivision_id, year]
        else:
            query = "SELECT * FROM groups WHERE subdivisionid = $1 ORDER BY name"
            params = [subdivision_id]
        
        async with self._get_connection(conn) as connection:
            rows = await connection.fetch(query, *params)
            return [Group(**dict(row)) for row in rows]
    
    async def get_with_stats(self, id: int, conn: Optional[Connection] = None) -> Optional[GroupWithStats]:
        """Получить группу со статистикой"""
        query = """
            SELECT 
                g.*,
                s.name as subdivision_name,
                COUNT(st.id) as students_count,
                COUNT(st.id) FILTER (WHERE st.isactive = true) as active_students_count,
                COUNT(st.id) FILTER (WHERE st.isbudget = true) as budget_students_count
            FROM groups g
            LEFT JOIN subdivisions s ON s.id = g.subdivisionid
            LEFT JOIN students st ON st.groupid = g.id
            WHERE g.id = $1
            GROUP BY g.id, s.name
        """
        
        async with self._get_connection(conn) as connection:
            row = await connection.fetchrow(query, id)
            return GroupWithStats(**dict(row)) if row else None
    
    async def get_all_with_stats(self, year: Optional[int] = None, conn: Optional[Connection] = None) -> List[GroupWithStats]:
        """Получить все группы со статистикой"""
        base_query = """
            SELECT 
                g.*,
                s.name as subdivision_name,
                COUNT(st.id) as students_count,
                COUNT(st.id) FILTER (WHERE st.isactive = true) as active_students_count,
                COUNT(st.id) FILTER (WHERE st.isbudget = true) as budget_students_count
            FROM groups g
            LEFT JOIN subdivisions s ON s.id = g.subdivisionid
            LEFT JOIN students st ON st.groupid = g.id
        """
        
        if year:
            query = base_query + " WHERE g.year = $1 GROUP BY g.id, s.name ORDER BY g.name"
            params = [year]
        else:
            query = base_query + " GROUP BY g.id, s.name ORDER BY g.name"
            params = []
        
        async with self._get_connection(conn) as connection:
            rows = await connection.fetch(query, *params)
            return [GroupWithStats(**dict(row)) for row in rows]
